Tetrahedral remeshing from C3T3

janetournois opened the pull request:
## Summary of Changes

This PR introduces a new version of `tetrahedral_isotropic_remeshing()` that takes a C3T3 (the output of Mesh_3) as input.

## Release Management

* Affected package(s): Tetrahedral_remeshing
* Feature/Small Feature (if any): [small feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Tetrahedral_remeshing_from_C3t3#Rationale), pre-approved by @lrineau, 2020/07/03
* Link to compiled documentation (obligatory for small feature) [ref manual](https://cgal.geometryfactory.com/~jtournois/Tetrahedral_remeshing_from_c3t3_v2/Tetrahedral_remeshing/group__PkgTetrahedralRemeshingRef.html#ga0197df917d3de1ddd75b3a5f35e54d85)
* License and copyright ownership: unchanged

**TODO**:
- [x] check branch size (@sloriot) (@lrineau: I have verified and it seems fine.)

MaelRL commented on Tetrahedral_remeshing/doc/Tetrahedral_remeshing/Tetrahedral_remeshing.txt:
```diff
@@ -50,6 +50,14 @@ Experimental evidence shows that a higher number of remeshing iterations
+Tetrahedral mesh, modified by our uniform tetrahedral remeshing method.
+(Left) Before remeshing, dihedral angles were in the interval [1.3; 177.8].
+(Right) After remeshing and keeping the same density,
+dihedral angles are the interval [12,7; 157.7].
```
```suggestion
dihedral angles are in the interval [12,7; 157.7].
```

MaelRL commented on Tetrahedral_remeshing/doc/Tetrahedral_remeshing/Tetrahedral_remeshing.txt:
```diff
@@ -97,5 +105,17 @@ setting the named parameter `remesh_boundaries` to `false`.
+
+\subsection ssecEx4 Tetrahedral Remeshing After Mesh Generation
+
+The tetrahedral remeshing algorithm is designed as a post-processing for
```
```suggestion
The tetrahedral remeshing algorithm is designed as a post-processing step for
```

janetournois replied:
why "step"? Imho "post-processing" is enough. Don't you think?
I would like to write "post-processing algorithm" but it would be too redundant

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
 
+/*!
+* \ingroup PkgTetrahedralRemeshingRef
+* remeshes a tetrahedral mesh.
```
```suggestion
* \brief Remeshes a tetrahedral mesh.
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
+*
+* This function takes as input a `Mesh_complex_3_in_triangulation_3`
+* and performs a sequence of atomic operations on the underlying triangulation
+* in order to generate as output a high quality mesh with a prescribed
```
```suggestion
* in order to generate as output a high-quality mesh with a prescribed
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
+* These atomic operations are performed as follows:
+*   - edge splits, until all edges satisfy a prescribed length criterion,
+*   - edge collapses, until all edges satisfy a prescribed length criterion,
+*   - edge flips, to locally improve dihedral angles, until they can
```
```suggestion
*   - edge flips, to locally improve dihedral angles until they can
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
+*   - global smoothing by vertex relocations,
+*   - re-projection of boundary vertices to the initial surface.
+*
+* This remeshing function can deal with multi-domains, boundaries, and features,
```
```suggestion
* This remeshing function can deal with multi-domains, boundaries, and features
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
+* Subdomains are defined by indices that
+* are stored in the cells of the input triangulation, following the `MeshCellBase_3`
+* concept.
+* The surfacic, linear and corner features (constrained edges and vertices)
```
```suggestion
* The surfacic, linear, and corner features (constrained edges and vertices)
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -256,6 +256,84 @@ void tetrahedral_isotropic_remeshing(
+*            the `Curve_index` type of the model of the `MeshDomainWithFeatures_3` concept used for mesh generation.
+* @tparam NamedParameters a sequence of \ref Remeshing_namedparameters "Named Parameters"
+*
+* @param c3t3 the complex containing the triangulation to be remeshed, of type `Tr`.
```
```suggestion
* @param c3t3 the complex containing the triangulation (of type `Tr`) to be remeshed.
```

MaelRL commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -380,9 +458,9 @@ void tetrahedral_isotropic_remeshing(
-  // perform remeshing
-  std::size_t nb_extra_iterations = 3;
-  remesher.remesh(max_it, nb_extra_iterations);
+    // perform remeshing
+    std::size_t nb_extra_iterations = 3;
+    remesher.remesh(max_it, nb_extra_iterations);
```
```suggestion
  // perform remeshing
  std::size_t nb_extra_iterations = 3;
  remesher.remesh(max_it, nb_extra_iterations);
```

lrineau commented on Tetrahedral_remeshing/examples/Tetrahedral_remeshing/mesh_and_remesh_polyhedral_domain_with_features.cpp:
```diff
@@ -0,0 +1,71 @@
+typedef CGAL::Surface_mesh<K::Point_3> Polyhedron;
+typedef CGAL::Polyhedral_mesh_domain_with_features_3<K, Polyhedron> Mesh_domain;
+
+#ifdef CGAL_CONCURRENT_MESH_3
+typedef CGAL::Parallel_tag Concurrency_tag;
+#else
+typedef CGAL::Sequential_tag Concurrency_tag;
+#endif
```
If you want to keep the possibility to mesh in parallel, then the `CMakeLists.txt` has to be modified.

lrineau commented on Tetrahedral_remeshing/examples/Tetrahedral_remeshing/tetrahedral_remeshing_example.cpp:
```diff
@@ -14,6 +14,22 @@ typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
 
 typedef CGAL::Tetrahedral_remeshing::Remeshing_triangulation_3<K> Remeshing_triangulation;
 
+bool generate_input(const std::size_t& n,
+                    const char* filename)
+{
+  typedef Remeshing_triangulation::Point Point;
+  Remeshing_triangulation tr;
+
+  CGAL::Random rng;
+  while (tr.number_of_vertices() < n)
+    tr.insert(Point(rng.get_double(-1., 1.), rng.get_double(-1., 1.), rng.get_double(-1., 1.)));
+
+  std::ofstream oFileT(filename, std::ios::out);
+  // writing file output;
+  oFileT << tr;
+
+  return (!oFileT.bad());
+}
```
I think that was added by error. That function is not used.
```suggestion
```

janetournois commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -252,10 +253,46 @@ void tetrahedral_isotropic_remeshing(
+* @todo finish this doc
+*/
+template<typename Tr, typename CornerIndex, typename CurveIndex>
+typename Tr::Tr_Base
```
@lrineau we have a pb here : `Tr_base` is not documented

MaelRL replied:
The triangulation type in the C3T3 is not really a template, it's a regular 3D triangulation of CGAL, so couldn't you just template the function with the `Vb` and `Cb` instead? Then you have:
```
Triangulation_3<Vb, Cb> convert(C3T3<Regular_triangulation_3<Cb, Vb>, ...)
```

lrineau replied:
There is a third parameter: the lock_ds:
```
template < class GT, class Tds_, class Lock_data_structure_ >
class Triangulation_3
```

janetournois replied:
That sounds good, thank you!

janetournois commented:
I just updated the link to the new reference manual at the top of this PR

lrineau commented on Tetrahedral_remeshing/examples/Tetrahedral_remeshing/mesh_and_remesh_polyhedral_domain_with_features.cpp:
```diff
@@ -0,0 +1,70 @@
+  // Mesh generation
+  C3t3 c3t3 = CGAL::make_mesh_3<C3t3>(domain, criteria);
+
+  Triangulation_3 tr = CGAL::convert_to_triangulation_3(c3t3);
+
```
```suggestion
  Triangulation_3 tr = CGAL::convert_to_triangulation_3(std::move(c3t3));

```

in my opinion, that is not the job of the CGAL reference manual to teach users about C++. But we can encourage good practice in our examples.

sloriot replied:
both options can be present (one being commented) and some comment should be added to make it clear what both options imply.

lrineau commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -252,10 +252,54 @@ void tetrahedral_isotropic_remeshing(
+/*!
+* Converts the triangulation referenced in the input to a `Triangulation_3`.
+* This function should be used to generate a valid triangulation
+* for tetrahedral remeshing remeshing, when the input triangulation is generated with the
```
```suggestion
* for tetrahedral remeshing, when the input triangulation is generated with the
```

lrineau commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -252,10 +252,54 @@ void tetrahedral_isotropic_remeshing(
 }
 
+/*!
+* Converts the triangulation referenced in the input to a `Triangulation_3`.
```
```suggestion
* Converts the triangulation contained in the input to a `Triangulation_3`.
```
"referenced" would mean that the c3t3 object detains a reference to an existing triangulation object, whereas the reality is that the triangulation is a (non-reference) data-member of the c3t3.

janetournois replied:
right! I thought it was a reference

lrineau commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -267,19 +269,24 @@ void tetrahedral_isotropic_remeshing(
-typename Tr::Tr_Base
-convert_to_base_triangulation(
+
+template<typename Tr,
+         typename CornerIndex,
+         typename CurveIndex>
+CGAL::Triangulation_3<typename Tr::Geom_traits,
+                      typename Tr::Triangulation_data_structure>
+convert_to_triangulation_3(
```
As I said, I prefer to mention the third template parameter: `Lock_data_structure`.

```suggestion
template<typename Tr,
         typename CornerIndex,
         typename CurveIndex>
CGAL::Triangulation_3<typename Tr::Geom_traits,
                      typename Tr::Triangulation_data_structure,
                      typename Tr::Lock_data_structure>
convert_to_triangulation_3(
```

lrineau commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -267,19 +269,24 @@ void tetrahedral_isotropic_remeshing(
-  using Tr_Base = typename Tr::Tr_Base;
-  Tr_Base tr;
+  using GT   = typename Tr::Geom_traits;
+  using TDS  = typename Tr::Triangulation_data_structure;
+
+  CGAL::Triangulation_3<GT, TDS> tr;
```
```suggestion
  using TDS  = typename Tr::Triangulation_data_structure;
  using SLDS = typename Tr::Lock_data_structure;

  CGAL::Triangulation_3<GT, TDS, SLDS> tr;
```

lrineau commented:
Note that this PR makes the Tetrahedral remeshing package depend on Mesh_3 even more.

janetournois commented:
That's true but this additionnal functionality is designed to make it work together with Mesh_3, so it should not be too much of an issue.

lrineau commented:
> That's true but this additionnal functionality is designed to make it work together with Mesh_3, so it should not be too much of an issue.

That is a topic that we will deal with later. See https://github.com/CGAL/cgal/issues/4810, planned for 5.2.

sloriot commented on Tetrahedral_remeshing/doc/Tetrahedral_remeshing/dependencies:
```diff
@@ -8,3 +8,4 @@ Triangulation_3
 BGL
 Mesh_3
 TDS_3
+Surface_mesh
```
why?

janetournois replied:
because one of the examples with Mesh_3 uses a `Surface_mesh` as input

sloriot commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -37,7 +37,7 @@ namespace CGAL
 /*!
 * \ingroup PkgTetrahedralRemeshingRef
-* remeshes a tetrahedral mesh.
+* Remeshes a tetrahedral mesh.
```
```suggestion
* remeshes a tetrahedral mesh.
```

sloriot commented on Tetrahedral_remeshing/include/CGAL/tetrahedral_remeshing.h:
```diff
@@ -252,10 +252,54 @@ void tetrahedral_isotropic_remeshing(
 
+/*!
+* \ingroup PkgTetrahedralRemeshingRef
+* Converts the triangulation contained in the input to a `Triangulation_3`.
```
```suggestion
* converts the triangulation contained in the input to a `Triangulation_3`.
```

lrineau commented:
@janetournois Can you please merge `master`, to solve the conflict?

janetournois commented:
Sure!

janetournois commented:
The testsuite shows again the warning : " warning STL4010: Various members of std::allocator are deprecated in C++17.", which had been fixed in the past with commit b6912d92eadfff1db25270cb8d3d134cd8f022fd ...

afabri commented:
Does silencing this warning really make sense?  Where does it come from? Because we have `-permissive-`  in this column of the testsuite?  Did you see that another warning gets silenced with `-D_SILENCE_CXX17_ADAPTOR_TYPEDEFS_DEPRECATION_WARNING` in the CXX_FLAGS?  So we do not deal with it consistently.

lrineau commented:
> Does silencing this warning really make sense? Where does it come from? Because we have `-permissive-` in this column of the testsuite? Did you see that another warning gets silenced with `-D_SILENCE_CXX17_ADAPTOR_TYPEDEFS_DEPRECATION_WARNING` in the CXX_FLAGS? So we do not deal with it consistently.


That was debated in another PR: https://github.com/CGAL/cgal/pull/4620#issuecomment-628433188, and we decided it was a bug from MSVC.

... Actually, @janetournois here the warning is a different one:
> warning STL4010: Various members of std::allocator are deprecated in C++17. Use std::allocator_traits instead of accessing these members directly. You can define _SILENCE_CXX17_OLD_ALLOCATOR_MEMBERS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning.

... and it has a different macro to define to hide it: `_SILENCE_CXX17_OLD_ALLOCATOR_MEMBERS_DEPRECATION_WARNING`.

All those warnings are from third-party dependencies (mostly from Boost). How should we deal with that?
  - Hide them only in our testsuite (using flags in our CXXFLAGS for the Microsoft platforms)?
  - Hide them globally in CGAL, with something in `<CGAL/config.h>`?

lrineau commented:
I pre-approuve.

lrineau commented on Tetrahedral_remeshing/test/Tetrahedral_remeshing/test_tetrahedral_remeshing_from_mesh_file.cpp:
```diff
@@ -0,0 +1,37 @@
+#include <CGAL/tetrahedral_remeshing.h>
+
+#include <CGAL/Tetrahedral_remeshing/tetrahedral_remeshing_io.h>
+#include "tetrahedral_remeshing_generate_input.h"
```
Missing file. See Travis error: https://travis-ci.com/github/CGAL/cgal/jobs/357279120#L3259

janetournois replied:
fixed. Sorry!

janetournois commented:
@lrineau I've been thinking about your question... I don't like the idea of hiding warnings only in the testsuite, because then the users would face them systematically and the testsuite would behave like it "does not care". Hiding them everywhere in CGAL seems to be a better/more consistent option.

janetournois commented:
@lrineau @maxGimeno @sloriot
I just added the macro `CGAL_CXX17` to fix the warnings/errors about using `std::not1` in Mesh_3 (deprecated in C++17, removed from C++20, but the replacing `std::not_fn` is introduced in C++17 so we need both).

Please just note that the macro `__cplusplus` is valid in MSVC since MSVC 2017 (see [here](https://docs.microsoft.com/fr-fr/cpp/build/reference/zc-cplusplus?view=vs-2019)) but requires to add `/Zc:__cplusplus` to the compiler options, so I added a test on `_MSVC_LANG` anyhow to keep as versatile as possible.

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-193.shtml.

lrineau commented:
Good, now that the PR is tested, @sloriot have you checked the branch size?

lrineau commented:
> Good, now that the PR is tested, @sloriot have you checked the branch size?

I verified myself and that seems fine.
